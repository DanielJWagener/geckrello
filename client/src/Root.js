// Simply creates a Redux store and passes it in to a provider tag.
// This logic was originally in index.js, but we've movied it here to
// make our testing process easier.
import React from "react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware, compose } from "redux";
import reduxThunk from "redux-thunk";
import reducers from "./reducers";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  reducers,
  {},
  composeEnhancers(applyMiddleware(reduxThunk))
);

export default ({ children }) => {
  return <Provider store={store}>{children}</Provider>;
};
