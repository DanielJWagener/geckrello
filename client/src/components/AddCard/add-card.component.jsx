import React from "react";
import { connect } from "react-redux";
import { addCard } from "../../actions";

import "./add-card.styles.scss";

export class AddCard extends React.Component {
  // This component has two modes: "prompt" and "input." The "prompt" mode is the initial mode, basically a button with a label like "add new card." When the user clicks that button, the mode changes to "input," and the user can input and submit a new card and title.
  state = {
    mode: "prompt",
    backgroundColor: "",
    cardTitle: "",
    borderColor: "#555"
  };

  // Hover effect (CSS's :hover pseudoselector wasn't creating the desired effect)
  backgroundColor = {
    default: "#e0e4ff",
    hover: "#9093ad"
  };

  hoverBackground = () => {
    this.setState({ backgroundColor: this.backgroundColor.hover });
  };

  resetBackground = () => {
    this.setState({ backgroundColor: this.backgroundColor.default });
  };

  // Set initial background color, get height for this.promptHeight
  addCardPromptRef = React.createRef();
  addCardInputRef = React.createRef();

  promptHeight = 0;
  inputHeight = 96; // TODO: don't hardcode this value

  componentDidMount() {
    this.setState({
      backgroundColor: this.backgroundColor.default
    });
  }

  // Mode and height toggle
  toggleMode = () => {
    if (this.state.mode === "prompt") {
      this.setState({ mode: "input" });
      this.resetBackground();
    } else {
      this.setState({ mode: "prompt" });
    }
  };

  // Card sumbit handler
  onFormSubmit = e => {
    e.preventDefault();

    // Validate input
    if (!this.state.cardTitle) {
      alert("Please enter a card title.");
      return false;
    }

    // Call addCard redux action if user has entered a card title
    if (this.state.cardTitle) {
      this.props.addCard(
        this.state.cardTitle,
        this.props.listHome,
        this.props.boardHome
      );

      // Clear input, change mode
      this.setState({ cardTitle: "", mode: "prompt", borderColor: "#555" });
    } else {
      return false;
    }
  };

  // Controlled input handler
  handleChange = e => {
    this.setState({ cardTitle: e.target.value });

    e.target.value
      ? this.setState({ borderColor: "#4fa644" })
      : this.setState({ borderColor: "red" });
  };

  render() {
    if (this.state.mode === "prompt") {
      return (
        <div
          className="add-card add-card__prompt"
          style={{ backgroundColor: this.state.backgroundColor }}
          onMouseOver={this.hoverBackground}
          onTouchStart={this.hoverBackground}
          onMouseLeave={this.resetBackground}
          onTouchMove={this.resetBackground}
          onClick={this.toggleMode}
          ref={this.addCardPromptRef}
        >
          <span className="add-card__prompt--plus">+</span>
          <span className="add-card__prompt--label">Add new card</span>
        </div>
      );
    } else {
      return (
        <div
          className="add-card add-card__form"
          style={{ backgroundColor: this.backgroundColor.default }}
          ref={this.addCardInputRef}
        >
          <form onSubmit={this.onFormSubmit} className="form">
            <div className="form__group">
              <input
                type="text"
                className="form form__input form__input--validation"
                value={this.state.cardTitle}
                onChange={this.handleChange}
                style={{ borderBottomColor: this.state.borderColor }}
                placeholder="New card title..."
                autoFocus
              />
            </div>
            <div className="form__group">
              <input type="submit" className="form__submit" value="Add Card" />
              <button className="form__cancel" onClick={this.toggleMode}>
                Cancel
              </button>
            </div>
          </form>
        </div>
      );
    }
  }
}

const mapStateToProps = ({ cards, board }) => {
  const boardHome = board ? board._id : "";
  return { boardHome, cards };
};

export default connect(mapStateToProps, { addCard })(AddCard);
