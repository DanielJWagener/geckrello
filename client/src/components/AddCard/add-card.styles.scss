@import "../../sass/abstracts/_variables.scss";

.list {
  width: 100%;
  background-color: $component-background;
  border-radius: 0.5rem;
  min-height: 120px;
  position: relative;
  margin-bottom: 2rem;
  z-index: 0;

  &__heading {
    font-size: 2rem;
    width: 90%;
    overflow: hidden;
    font-family: sans-serif;
    padding: 1rem 1rem 1.5rem 1rem;
  }

  &__remove {
    position: absolute;
    top: 3px;
    right: 1rem;
    cursor: pointer;
    font-size: 3rem;

    &:hover {
      color: red;
    }
  }

  &__cards {
    padding: 0.5rem 0.5rem 0.5rem 0.5rem;
  }

  &__addCard {
    border-radius: 0 0 0.5rem 0.5rem;
    font-family: sans-serif;

    &-prompt {
      padding: 1rem 1rem 1.2rem 1rem;

      overflow: hidden;
      display: flex;
      justify-content: flex-start;
      align-items: center;

      -webkit-user-select: none;
      user-select: none;
      cursor: pointer;

      &:link,
      &:visited {
        color: inherit;
        text-decoration: none;
      }

      &--plus {
        font-size: 4rem;
        margin-right: 1rem;
      }

      &--label {
        font-size: 2rem;
      }
    }

    &-form {
      width: 100%;
    }
  }

  &:after {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    z-index: -1;
    box-shadow: 0 1.5rem 4rem rgba(0, 0, 0, 0.3);
  }
}

.form {
  padding: 0.5rem;

  &__group:not(:last-child) {
    margin-bottom: 0.5rem;
  }

  &__input {
    display: block;
    width: 100%;
    font-size: 2rem;
    line-height: 3.5rem;
    padding: 0 1rem;
    border-radius: 0.5rem;
    border: none;
    color: #333;

    &--validation {
      border-bottom-style: solid;
      border-bottom-width: 3px;
      border-bottom-color: #555;
    }
  }

  &__submit,
  &__edit {
    display: inline-block;
    padding: 1rem 1.5rem;
    border-radius: 0.5rem;
    border: none;
    font-size: 2rem;
    background-color: $color-secondary;
    color: white;
    cursor: pointer;
    transition: all 0.2s;

    &:focus,
    &:active {
      outline: none;
      transform: translateY(2px);
    }

    &:hover {
      background-color: $color-secondary-dark;
    }
  }

  &__cancel {
    display: inline-block;
    margin-left: 1rem;
    font-size: 1.5rem;
    text-decoration: none;
    color: #333;
    background-color: transparent;
    border: none;
    cursor: pointer;

    &:hover {
      color: red;
    }
  }
}
