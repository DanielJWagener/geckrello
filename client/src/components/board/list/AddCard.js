import React from "react";
import { connect } from "react-redux";
import { addCard } from "../../../actions";

class AddCard extends React.Component {
  // This component has two modes: "prompt" and "input." The "prompt" mode is the initial mode, basically a button with a label like "add new card." When the user clicks that button, the mode changes to "input," and the user can input and submit a new card and title.
  state = {
    mode: "prompt",
    backgroundColor: "",
    cardTitle: "",
    borderColor: "#555"
  };

  // Hover effect (CSS's :hover pseudoselector wasn't creating the desired effect)
  backgroundColor = {
    default: "#e0e4ff",
    hover: "#9093ad"
  };

  hoverBackground = () => {
    this.setState({ backgroundColor: this.backgroundColor.hover });
  };

  resetBackground = () => {
    this.setState({ backgroundColor: this.backgroundColor.default });
  };

  // Set initial background color, get height for this.promptHeight
  addCardPromptRef = React.createRef();
  addCardInputRef = React.createRef();

  promptHeight = 0;
  inputHeight = 96; // TODO: don't hardcode this value

  componentDidMount() {
    this.setState({
      backgroundColor: this.backgroundColor.default
    });

    this.promptHeight = this.addCardPromptRef.current.clientHeight;
  }

  // Mode and height toggle
  toggleMode = () => {
    if (this.state.mode === "prompt") {
      this.setState({ mode: "input" });
    } else {
      this.setState({ mode: "prompt" });
    }

    let currentHeight =
      this.state.mode === "input" ? this.promptHeight : this.inputHeight;
    this.props.setSpansUpdate(currentHeight);
  };

  // Card sumbit handler
  onFormSubmit = e => {
    e.preventDefault();

    // Validate input
    if (!this.state.cardTitle) {
      alert("Please enter a card title.");
      return false;
    }

    // Generate a unique key and cardId in the event of duplicate card titles
    const generateCardId = inputId => {
      if (!this.props.cards.find(card => card.cardId === inputId)) {
        return inputId;
      } else {
        inputId = `${inputId}-duplicate`;
        return generateCardId(inputId);
      }
    };
    let cardId = generateCardId(this.state.cardTitle);

    // Call addCard redux action if user has entered a card title
    if (this.state.cardTitle) {
      this.props.addCard(this.state.cardTitle, cardId, this.props.listHome);

      // Clear input, change mode
      this.setState({ cardTitle: "", mode: "prompt", borderColor: "#555" });
      this.props.setSpansUpdateForCard(this.promptHeight);
    } else {
      return false;
    }
  };

  // Controlled input handler
  handleChange = e => {
    this.setState({ cardTitle: e.target.value });

    e.target.value
      ? this.setState({ borderColor: "#4fa644" })
      : this.setState({ borderColor: "red" });
  };

  render() {
    if (this.state.mode === "prompt") {
      return (
        <div
          className="list__addCard list__addCard-prompt"
          style={{ backgroundColor: this.state.backgroundColor }}
          onMouseOver={this.hoverBackground}
          onMouseLeave={this.resetBackground}
          onClick={this.toggleMode}
          ref={this.addCardPromptRef}
        >
          <span className="list__addCard-prompt--plus">+</span>
          <span className="list__addCard-prompt--label">Add new card</span>
        </div>
      );
    } else {
      return (
        <div
          className="list__addCard list__addCard-form"
          style={{ backgroundColor: this.backgroundColor.default }}
          ref={this.addCardInputRef}
        >
          <form onSubmit={this.onFormSubmit} className="form">
            <div className="form__group">
              <input
                type="text"
                className="form form__input form__input--validation"
                value={this.state.cardTitle}
                onChange={this.handleChange}
                style={{ borderBottomColor: this.state.borderColor }}
                placeholder="New card title..."
                autoFocus
              />
            </div>
            <div className="form__group">
              <input type="submit" className="form__submit" value="Add Card" />
              <button className="form__cancel" onClick={this.toggleMode}>
                Cancel
              </button>
            </div>
          </form>
        </div>
      );
    }
  }
}

const mapStateToProps = state => {
  return { cards: state.cards };
};

export default connect(
  mapStateToProps,
  { addCard }
)(AddCard);
