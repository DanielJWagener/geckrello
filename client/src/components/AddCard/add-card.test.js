import { shallow } from "enzyme";
import React from "react";
import AddCard from "./add-card.component";
import Root from "../../Root";

it("renders AddCard component", () => {
  expect(
    shallow(
      <Root>
        <AddCard />
      </Root>
    )
  ).toMatchSnapshot();
});
