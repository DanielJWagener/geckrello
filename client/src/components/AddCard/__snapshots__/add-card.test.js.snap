// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AddCard component 1`] = `ShallowWrapper {}`;
